### PR TITLE
#5592: Enable optimized prefill for Falcon7b

### PR DESCRIPTION
**Pull request by s-jovic**
Enable optimized prefill for 128, 1024 and 2048 seq len by default on wormhole B0 (GS is kept as is).

Previously it has been disabled by default since there was some e2e perf degradations for 128 seq len. This PR resolves this issues, enables optimized prefill by default and bumps up matmul subblocks now that the matmul hangs have been resolved (https://github.com/tenstorrent/tt-metal/issues/7066).

The remaining cleanup/fixes is listed in https://github.com/tenstorrent/tt-metal/issues/8349 and will be tackled next.

- Post commit: https://github.com/tenstorrent/tt-metal/actions/runs/9031161202

### CODE REVIEW

**Review by pavlejosipovic — APPROVED** [review]

**Comment by s-jovic** [comment]
Nightly fd dispatch: https://github.com/tenstorrent/tt-metal/actions/runs/9064660786
Model perf multi chip: https://github.com/tenstorrent/tt-metal/actions/runs/9063195161

Model perf single chip: https://github.com/tenstorrent/tt-metal/actions/runs/9064551488
- this one has been stopped as mamba tests hung; falcon tests executed fine, but [rerunning the pipeline here](https://github.com/tenstorrent/tt-metal/actions/runs/9065710709) to double check what happens with mamba

@tt-rkim can you give the review? I will see in the morning what happened with model perf pipeline, I won't check in the change if the problems are still there of course

File: models/demos/falcon7b/tests/test_perf_falcon.py [hunk 1]
```diff
@@ -507,10 +505,9 @@ def run_perf_wh_bare_metal(
...
-            ("prefill", 32, 1, 128, 0, "BFLOAT16-DRAM", 0.97, 0.99, 0.96, 0.17),
-            ("prefill", 32, 1, 128, 0, "BFLOAT16-L1", 0.97, 0.99, 0.96, 0.17),
-            ("prefill", 32, 1, 256, 0, "BFLOAT16-DRAM", 0.98, 0.99, 0.96, 0.2),
-            ("prefill", 32, 1, 256, 0, "BFLOAT16-L1", 0.98, 0.99, 0.96, 0.2),
+            ("prefill", 32, 1, 128, 0, "BFLOAT16-L1_SHARDED", 0.97, 0.99, 0.96, 0.1),
+            # ("prefill", 32, 1, 1024, 0, "BFLOAT16-L1_SHARDED", 0.98, 0.99, 0.96, 1), Enable when #8349 is resolved
+            # ("prefill", 32, 1, 2048, 0, "BFLOAT16-L1_SHARDED", 0.99, 0.99, 0.97, 1),
             ("decode", 32, 32, 1, 128, "BFLOAT16-DRAM", 0.91, 0.92, 0.93, 0.15),
             ("decode", 32, 32, 1, 128, "BFLOAT16-L1", 0.91, 0.92, 0.93, 0.15),
             ("decode", 32, 32, 1, 128, "BFLOAT16-L1_SHARDED", 0.92, 0.95, 0.95, 0.1),
```
Comment (skhorasganiTT) [inline models/demos/falcon7b/tests/test_perf_falcon.py:513]: The Falcon7b demo currently uses the BFLOAT16-DRAM config for grayskull/wormhole/t3000. We should either modify the config used in the demo or keep the dram config tested here. 
Also, I see that in the tests seq_len > 1024 is being skipped, while that is the default seq len in the demo. If only the dram config supports that seq len, then we should probably keep using that config.

File: models/demos/falcon7b/tests/test_perf_falcon.py [hunk 2]
```diff
@@ -469,16 +469,14 @@ def run_perf_wh_bare_metal(
     ):
         if model_config_str == "BFLOAT16-L1_SHARDED" and kv_cache_len == 2047:
             pytest.skip(f"kv_cache_len={kv_cache_len} does not fit with L1_SHARDED")
-        if model_config_str == "BFLOAT16-L1_SHARDED" and llm_mode == "prefill":
-            pytest.skip(f"prefill does not support L1_SHARDED")
```
Comment (skhorasganiTT) [inline models/demos/falcon7b/tests/test_perf_falcon.py:473]: The l1 sharded config was originally designed as the optimized decode config, should we be using that for prefill? Same question for all the other tests that have been modified to use l1 sharded for prefill.

File: models/demos/falcon7b/tests/test_falcon_end_to_end.py [hunk 3]
```diff
@@ -294,81 +295,54 @@ def run_test_FalconCausalLM_end_to_end(
...
-@pytest.mark.parametrize(
-    "num_layers, out_pcc, cache_pcc",
-    ((2, 0.98, 0.98), (32, 0.86, 0.86)),
-    ids=["layers_2", "layers_32"],
+    ids=[
+        "prefill_seq32_bf16_l1",
+        "prefill_seq32_bf16_dram",
+        "prefill_seq128_bf16_l1",
+        "prefill_seq1024_bf16_l1",
+        "prefill_seq2048_bf16_l1",
```
Comment (skhorasganiTT) [inline models/demos/falcon7b/tests/test_falcon_end_to_end.py:315]: Can these be renamed to l1sharded if that is the config that is supposed to be used?

File: models/demos/falcon7b/tests/test_falcon_end_to_end.py [hunk 4]
```diff
@@ -294,81 +295,54 @@ def run_test_FalconCausalLM_end_to_end(
...
-
-    if (
-        is_grayskull
-        and model_config_str == "BFLOAT16-L1"
-        and num_layers == 32
-        and llm_mode == "prefill"
-        and seq_len == 1024
-    ):
-        pytest.skip("#7933: Out of DRAM space error for tensor")
-
```
Comment (skhorasganiTT) [inline models/demos/falcon7b/tests/test_falcon_end_to_end.py:371]: @tt-rkim recently added these changes to resolve some pipeline failures, but I am not sure if we are running this test in pipelines anymore. Also this test is essentially a duplicate of test_perf_falcon.py, is it safe to delete it entirely?

File: models/demos/falcon7b/tt/falcon_causallm.py [hunk 5]
```diff
@@ -45,12 +40,14 @@ def __init__(
...
 
+        self.is_grayskull = is_grayskull()
+
         lm_head_weight = None
         if self.state_dict:
             lm_head_weight = self.state_dict["lm_head.weight"]
             lm_head_weight = torch.transpose(lm_head_weight, -2, -1)
 
-        if self.seq_len > 512:
+        if not self.is_grayskull and self.seq_len > 512:
```
Comment (skhorasganiTT) [inline models/demos/falcon7b/tt/falcon_causallm.py:50]: Would it be better to keep a OPTIMIZED_MODE_PREFILL setting in model_config and set that to is_grayskull()? If all the is_grayskull calls in the model files are coupled, maybe we should only set it once to avoid confusion in the future.

File: tests/scripts/run_performance.sh [hunk 6]
```diff
@@ -36,7 +36,7 @@ run_perf_models_llm_javelin() {
     local tt_arch=$1
     local test_marker=$2
 
-    env pytest models/demos/falcon7b/tests -m $test_marker
+    env WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml pytest models/demos/falcon7b/tests -m $test_marker
 
     if [ "$tt_arch" == "wormhole_b0" ]; then
         env pytest models/demos/mamba/tests -m $test_marker
```
Comment (skhorasganiTT) [inline tests/scripts/run_performance.sh:42]: Have you tested `WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml` with the [t3000](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/t3000/falcon7b) and [wormhole](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/wormhole/falcon7b) demos? If there is no drop in e2e perf when setting [perf_mode](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/t3000/falcon7b/demo_t3000.py#L10)=True, could you please update the command in the readme to use this?

**Review by djordje-tt — APPROVED** [review]

File: tests/scripts/run_performance.sh [hunk 7]
```diff
@@ -36,12 +36,12 @@ run_perf_models_llm_javelin() {
     local tt_arch=$1
     local test_marker=$2
 
-    env pytest models/demos/falcon7b/tests -m $test_marker
-
```
Comment (skhorasganiTT) [inline tests/scripts/run_performance.sh:40]: Why was this moved to later in the pipeline?

Comment (s-jovic) [reply tests/scripts/run_performance.sh:40]: For testing purposes, I am seeing some hangs after our tests

**Comment by s-jovic** [comment]
@skhorasganiTT just a note, there is a hang in Mamba test executing after our Falcon tests that I am having trouble reproducing locally, so the PR is blocked atm. Once this is resolved I will address the comments.

**Comment by skhorasganiTT** [comment]
> @skhorasganiTT just a note, there is a hang in Mamba test executing after our Falcon tests that I am having trouble reproducing locally, so the PR is blocked atm. Once this is resolved I will address the comments.

@s-jovic You can comment out [these lines](https://github.com/tenstorrent/tt-metal/blob/main/tests/scripts/run_performance.sh#L36-L38) for the mamba test and re-run the pipeline locally or on CI to verify the changes (and uncomment once finished)

File: models/demos/falcon7b/tests/test_perf_falcon.py (hunk 2 above)
Comment (s-jovic) [reply models/demos/falcon7b/tests/test_perf_falcon.py]: I reverted back to DRAM to avoid introducing confusion; We actually have one mode only for optimized prefill, so we will clean this up afterwards.

File: models/demos/falcon7b/tests/test_falcon_end_to_end.py (hunk 4 above)
Comment (s-jovic) [reply models/demos/falcon7b/tests/test_falcon_end_to_end.py:371]: I'd keep both of them - shouldn't perf falcon actually run some iterations before measuring time for warm-up? That's how we do it on Falcon40b for example.

**Review by skhorasganiTT — COMMENTED** [review]
Thanks for making all the fixes @s-jovic. As we were discussing offline could you also please disable optimized prefill in the gs/wh/t3000 demos since they use seq len 1024 (if it does not work) and verify they still work (with/without perf mode)?

**Comment by s-jovic** [comment]
@skhorasganiTT disabled optimized attention since that is the source of the segfault. Tried the demos, GS and WH work as expected, T3000 demo doesn't raise any errors, but it seems to get into some loop for one of the users - was kinda hard to understand which one as the terminal changes quickly with 256 users; however I see this running on main as well so I don't think it is a consequence of this PR.

@tt-rkim 

Model perf: https://github.com/tenstorrent/tt-metal/actions/runs/9129591265 (branch that is created from this one just has commented out Mamba test; GS fails on main)
Nightly dispatch: https://github.com/tenstorrent/tt-metal/actions/runs/9129591265 (GS fails on main also)
Multi device model perf: https://github.com/tenstorrent/tt-metal/actions/runs/9063195161 (from a couple of days ago, no changes made affecting this and the t3k pipelines were being under some refactoring these days)

**Review by skhorasganiTT — APPROVED** [review]

**Comment by s-jovic** [comment]
Pipelines after the latest rebase that includes T3K test refactoring:

Nightly: https://github.com/tenstorrent/tt-metal/actions/runs/9156334554 (GS error that exists on main as well) 
Model perf: https://github.com/tenstorrent/tt-metal/actions/runs/9156352472 (GS error that exists on main as well)
T3K model perf:  https://github.com/tenstorrent/tt-metal/actions/runs/9156354009

**Review by tapspatel — APPROVED** [review]
t3000 files look good

**Review by tt-rkim — APPROVED** [review]
